src/server/main_service.cc

```diff
@@ -1141,6 +1148,31 @@ void Service::DispatchCommand(CmdArgList args, facade::ConnectionContext* cntx)
     dfly_cntx->reply_builder()->CloseConnection();
   }
 
+  // if this is a read command, and client tracking has enabled,
+  // start tracking updates to the keys in this read command
+  // notify the client when there is update, see PostUpdate() in db_slice.cc
+  if ((cid->opt_mask() & CO::READONLY) && dfly_cntx->conn()->IsTrackingOn()) {
+    OpResult<KeyIndex> key_index_res = DetermineKeys(cid, args_no_cmd);
+    if (!key_index_res)
+      return (*cntx)->SendError(key_index_res.status());
+
+    const auto& key_index = *key_index_res;
+    vector<string_view> keys_to_track;
+    for (unsigned i = key_index.start; i < key_index.end; i += key_index.step) {
+      string_view key = ArgS(args_no_cmd, i);
+      keys_to_track.push_back(key);
+    }
+
+    // let's pass thread id and connection to db_slice for tracking
+    int32_t tid = util::ProactorBase::GetIndex();
+
+    // uint32_t client_id = dfly_cntx->conn()->GetClientId();
+    auto cb = [&](Transaction* t, EngineShard* shard) {
+      return OpTrackKeys(t->GetOpArgs(shard), dfly_cntx, tid, keys_to_track, args);
+    };
+    dfly_cntx->transaction->ScheduleSingleHopT(cb);
+  }
+
```

Comment:
- We currently don't allow simply re-using the transaction 🤔 Does it work reliably? 
- We could be doing this from the transaction when it's concluding, so we don't need to perform a hop for this

Reply:
Also commands that end with STORE have a bonus key: `key_index.bonus`

Reply:
now i can say it doesn't work reliably especially when multi dispatch is running... and thanks again for the `Refurbish()` function. That solves the issue.

Reply:
agreed, fixed!